### ISSUE TITLE
Build Python packages using the limited API

### ISSUE THREAD

**Issue by vyasr** [rapidsai/build-planning#42]
Python has [a limited API](https://docs.python.org/3/c-api/stable.html) that is guaranteed to be stable across minor releases. Any code using the Python C API that limits itself to using code in the limited API is guaranteed to also compile on future minor versions of Python within the same major family. More importantly, all symbols in the current (and some historical) version of the limited API are part of Python's stable ABI, which also does not change between Python minor versions and allows extensions compiled against one Python version to continue working on future versions of Python.

Currently RAPIDS builds a single wheel per Python version. If we were to compile using the Python stable ABI, we would be able to instead build a single wheel that works for all Python versions that we support. There would be a number of benefits here:
- **Reduced build time**: This benefit is largely reduced by #33, since if we build the C++ components as standalone wheels they are already Python-independent (except when we actually use the Python C API in our own C libraries; the only example that I'm currently aware of in RAPIDS is ucxx). The Python components alone are generally small and easy to build. We'll still benefit, but the benefits will be much smaller.
- **Reduced testing time**: Currently we run test across a number of Python versions for our packages on every PR. We often struggle with what versions need to be tested each time. If we were to only build a single wheel that runs on all Python versions, it would be much easier to justify a consistent strategy of always testing e.g. the earliest and latest Python versions. We may still want to test more broadly in nightlies, but really the only failure mode here is if a patch release is made for a Python version that is neither the earliest nor the latest, _and_ that patch release contains breaking changes. That is certainly possible (e.g. [the recent dask failure that forced us to make a last-minute patch](https://github.com/dask/dask/pull/11035)), but it's infrequent enough that we don't need to be testing regularly.
- **Wider support matrix**: Since we'll have a single binary that works for all Python versions, maintaining the full support matrix will be a lot easier and we won't feel as much pressure to drop earlier versions in order to support newer ones.
- **Day 0 support**: Our wheels will work for new Python versions as soon as they're released. Of course, if there are breaking changes then we'll have to address those, but in the average case where things do work users won't be stuck waiting on us.
- **Better installation experience**: Having a wheel that automatically works across Python versions will reduce the frequency of issues that are raised around our pip installs.

Here are the tasks (some ours, some external) that need to be accomplished to make this possible:
- **Making Cython compatible with the limited API**: [Cython has preliminary support for the limited API](https://cython.readthedocs.io/en/latest/src/changes.html#initial-support-for-limited-api). However, this support is still experimental, and most code still won't compile. I have been making improvements to Cython itself to fix this, and I now have a local development branch of Cython where I can compile most of RAPIDS (with additional changes to RAPIDS libraries). We won't be able to move forward with releasing production abi3 wheels until this support in Cython is released. This is going to be the biggest bottleneck for us.
- **nanobind support for the limited API**: nanobind can already produce abi3 wheels when compiled with Python 3.12 or later. Right now we use nanobind in pylibcugraphops, and nowhere else.
- **Removing C API usage in our code**: RAPIDS makes very minimal direct usage of the Python C API. The predominant use case that I see is creating memoryviews in order to access some buffers directly. We can fix this by constructing buffers directly. The other thing we'll want to do is remove usage of the NumPy C API, which has no promise of supporting the limited API AFAIK. That will be addressed in #41. Other use cases can be addressed incrementally. 
- **Intermediate vs. long-term**: If Cython support for the limited API ends up being released before RAPIDS drops support for Python 3.10, we may be in an intermediate state where we still need to build a version-specific wheel for 3.10 while building an abi3 wheel for 3.11+ (and 3.12+ for pylibcugraphops due to nanobind). If that is the case, it shouldn't cause much difficulty since it'll just involve adding a tiny bit of logic on top of our existing GH workflows.

**Comment by jakirkham** [comment]
It is worth noting that [the Python Buffer Protocol C API landed in Python 3.11]( https://docs.python.org/3/whatsnew/3.11.html#whatsnew311-c-api-new-features ) (additional [ref]( https://github.com/python/cpython/blob/d0f1307580a69372611d27b04bbf2551dc85a1ef/Misc/stable_abi.toml#L2273 )). So think that is a minimum for us

Also find [this listing of functions in the Limited and Stable API]( https://github.com/python/cpython/blob/d0f1307580a69372611d27b04bbf2551dc85a1ef/Misc/stable_abi.toml ) quite helpful

**Comment by vyasr** [comment]
Yes. I have been able to build most of RAPIDS using Cython's limited API supported (along with some additional changes I have locally) in Python 3.11. Python 3.11 is definitely a must. But as I said above in the "intermediate vs long-term" bullet, we could still benefit before dropping Python<3.11 support by building one wheel for each older Python version and then build an abi3 wheel to be used for Python 3.11+.

**Comment by vyasr** [comment]
I've made PRs ro rmm, raft, and cuml that address the issues in those repos. I've also taken steps to remove ucxx's usage of the the numpy C API (#41), which in turn removes one of its primary incompatibilities. The last major issue in RAPIDS code that I see is the usage of the array module in the Array class that is vendored by both kvikio and ucxx (and ucx-py). If that can be removed, then I think we'll be in good shape on the RAPIDS end, and we'll just be waiting on support for this feature in Cython itself. @jakirkham expressed interest in helping out with that in the process of making that Array class more broadly usable.

**Comment by da-woods** [comment]
A small warning here:

There's definitely places where Cython is substituting private C API for private Python API, so future compatibility definitely isn't guaranteed (it'll just be a runtime failure rather than a compile-time failure). We'll see how that evolves - I hope to be able to make some of these warnings rather than failures (since it's largely just non-essential introspection support).

We're also having to build a few more runtime version-checks into our code. Which is obviously a little risky because although you're compiling the same thing, you're taking different paths on different Python versions.

So the upshot is that your testing matrix probably doesn't reduce to a single version. (From Cython's point of view the testing matrix probably expands, because we really should be testing combinations like `Py_LIMITED_API=0x03090000` with Python 3.12 and that gets big quite quickly so I don't know how we're going to do that)

**Comment by vyasr** [comment]
Thanks for chiming in here @da-woods! I appreciate your comments. I agree that there is more complexity around testing here than simply a set and forget single version. At present, RAPIDS typically supports 2 or 3 Python versions at a time. We tend to lag a bit behind NEP 29/SPEC 0 timelines, so we support older versions a bit longer at the expense of not supporting new ones until they've been out for a bit. A significant part of the resource constraint equation for us is certainly on the testing side since running our full test suites on multiple Python versions adds up quickly. The way that I had envisioned this working, if we did move forward, would be that we built on the oldest supported Python (e.g. `Py_LIMITED_API=0x03090000`) and then we ran tests on the earliest and latest Python we supported (e.g. 3.9 and 3.11). The big benefit of using the limited API in this space would be that we could bump up the latest supported Python version without needing to move the earliest. The assumption would be that by the time a new Python version was released (e.g. 3.12), we would have gone through enough patch release of the previous release (3.11) to trust that nothing would be breaking in future patch releases. Of course, in practice that's probably not true: CPython certainly doesn't always strictly follow SemVer rules for patch releases, and to be fair Hyrum's law certainly applies to a project at that scale. Beyond that Cython's use of CPython internals certainly means that we could be broken even by patch releases. In practice what this would probably mean is that we would run tests as mentioned above on a frequent basis (on every PR), then run a larger test matrix infrequently (say, nightly or weekly). IOW even with limited API builds we would definitely still want to do broader testing to ensure that such builds are actually as compatible as they claim to be. However, I'd hope that the scale of that testing would be reduced.

**Comment by vyasr** [comment]
With the latest versions of branch-24.10, which contain a number of changes I made over the past few months for limited API compatibility along with the removal of pyarrow and numpy as build requirements in cudf and cuspatial, most of RAPIDS now builds with the limited API flag on. I have run some smoke tests and things generally work OK, but I haven't done anything extensive. ucxx and kvikio remain outstanding since we need to rewrite the Array class to not use the Python array's C API since that does not support the limited API. The latest tests can be seen in https://github.com/rapidsai/devcontainers/pull/278.

**Comment by da-woods** [comment]
I don't know if it's any help, but the quickest non-`array.array` API way I've found to allocate memory is:

```
cdef mview_cast = memoryview.cast
cdef Py_ssize_t[::1] new_Py_ssize_t_array(Py_ssize_t n):
    return mview_cast(PyMemoryView_FromObject(PyByteArray_FromStringAndSize(NULL, n* sizeof(Py_ssize_t))), "q")
```

It isn't as good, but it's surprisingly close given how much it actually does. "Only" 70% slower.

You probably have to replace

```
mv = PyMemoryView_FromObject(obj)
pybuf = PyMemoryView_GET_BUFFER(mv)
```

with `PyObject_GetBuffer` and an appropriate `PyBuffer_Release` in the destructor. But you probably should be doing that anyway - you're currently keeping a pointer to the data while not retaining a buffer-reference. That means things like `bytearray` could potentially be resized from under you.

**Comment by vyasr** [comment]
Thanks for the tip David! That could be helpful, but I'll have to look at the Array class more closely to be sure. I suspect that there are larger refactorings of our code base that could be done to make this unnecessary.

**Comment by jakirkham** [comment]
Had played with a few different approaches to rewrite `new_Py_ssize_t_array` including David's `PyByteArray_FromStringAndSize` suggestion. In the end found that simply using [Cython's own `array` type]( https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#cython-arrays ) worked reasonably well without compromising too much on performance or clarity. So we have rewritten `new_Py_ssize_t_array` accordingly

That should address this piece of this issue. Though am sure there are other things we may still need to do

**Comment by vyasr** [comment]
Wow that's awesome. Thanks for trying that out John! I'll refresh https://github.com/rapidsai/devcontainers/pull/278 for ucxx and kvikio and see what breaks next.

**Comment by vyasr** [comment]
Looks like ucxx is using a few functions that aren't part of the limited API. I'll look into updating those the next time I have a bit of free time.

**Comment by jameslamb** [comment]
> _Looks like ucxx is using a few functions that aren't part of the limited API (`PyUnicode_1BYTE_DATA`, `PyObject_CallMethodObjArgs`)_

Linking some background on that that might be helpful: https://github.com/rapidsai/ucxx/pull/276#issuecomment-2338028455

**Comment by jakirkham** [comment]
Looking at [the CPython API docs for `PyObject_CallMethodObjArgs`]( https://docs.python.org/3/c-api/call.html#c.PyObject_CallMethodObjArgs ), it says it is part of the Stable ABI

That said the PR James linked to uses `PyObject_CallMethodOneArg`, which appears to be absent from the Stable ABI. Looking at [the implementation of `PyObject_CallMethodOneArg`]( https://github.com/python/cpython/blob/85799f1ffd5f285ef93a608b0aaf6acbb464ff9d/Include/cpython/abstract.h#L68-L75 ) this is mainly a convenience wrapper around `PyObject_VectorcallMethod`, which is part of the Stable ABI. So it should be possible to rewrite using the Stable ABI

The `PyUnicode_1BYTE_DATA` appears to be more complicated as much of Python's C API for Unicode Objects is not part of the Stable ABI. It might be worth talking to Peter to better understand what we would like to accomplish in that code. For example if we just needs some C strings to create that error message, we might be better off setting them globally and using them directly in [that error construction call]( https://github.com/rapidsai/ucxx/blob/b94958dff044a1c063beff0d998e402525f913ca/cpp/python/src/future.cpp#L98-L101 )

**Comment by jakirkham** [comment]
It is worth noting that Python 3.13's free-threading build (GIL disabled) is not compatible with the Limited C API or Stable ABI

Of course the non-free-threading (GIL enabled) Python 3.13 can use the Limited C API and Stable ABI

Ref: https://docs.python.org/3/howto/free-threading-extensions.html#building-extensions-for-the-free-threaded-build

**Comment by vyasr** [comment]
Yeah I'm not thinking too much about the free-threading builds yet. Given the phrasing:

> The free-threaded build does not **currently** support the [Limited C API](https://docs.python.org/3/c-api/stable.html#limited-c-api) or the stable ABI.

(emphasis mine). It seems to me like they can't make free threaded builds the default without something akin to a major version bump unless they get the limited/stable ABI working, because without that all of a sudden it would be valid to install a bunch of incompatible packages into an environment. Either that, or they try and get installers pip/uv/etc to handle this, but even then you'll see tons of errors from people with older versions of those installers in environments that don't get updated. For the moment I'm OK waiting to see what direction this goes, but I'd personally be pretty surprised if they decided to break the limited API altogether to get free threading in as default.

**Comment by jakirkham** [comment]
The reason I mention it is most of the value of the Stable ABI and Limited API is building a package for one Python version and reusing that package for multiple versions. However with this change in Python 3.13 and the fact that the Python Buffer Protocol was only added in Python 3.11, it means we can only confidently build something for Python 3.11 and allow it to be installed for Python 3.12.

Before Python 3.11 we lack the Python Buffer Protocol, so have to build per Python version there.

Starting with Python 3.13, we lack an ability to constrain a package to only non-free-threading builds. IOW we have to assume a package using the Stable ABI and Limited API could wind up being installed on a free-threading build of Python unexpectedly. So we would need to build for Python 3.13 separately (and if we want to build for free-threading we would need to add that as an additional build).

IOW we are stuck with an island of compatibility at the moment with Python 3.11 and 3.12 on the Stable ABI and Limited API.

**Comment by vyasr** [comment]
> Starting with Python 3.13, we lack an ability to constrain a package to only non-free-threading builds. IOW we have to assume a package using the Stable ABI and Limited API could wind up being installed on a free-threading build of Python unexpectedly. 

This can't be true... The page you linked above says:

> C API extensions need to be built specifically for the free-threaded build. The wheels, shared libraries, and binaries are indicated by a t suffix.

Installers have to respect this. If they don't, free-threaded builds would start breaking all over the places when incompatible extensions are installed. My assumption is that for as long as free-threaded builds are not the default build, if we want to support them we will have to build separate wheels for them, and if that is the case we could use limited API builds for the default builds and then separate builds for each Python version under free threading.

I don't think anyone knows what will happen at the point when free-threading becomes the default yet, though.

**Comment by vyasr** [comment]
I don't understand what you mean by "there is not a way to separate them".

> The wheels, shared libraries, and binaries are indicated by a t suffix.

The package files are delineated, so as long as you appropriately tag wheels that are built with the limited API and installers are made aware of this too, what more do we need?

**Comment by jakirkham** [comment]
Could you please outline a proposal of how you see this being used?

How will we build for/support/package RAPIDS for each Python version in the range 3.10-3.13 (including free-threading)?

Think that will make it easier to identify and discuss edge cases

**Comment by vyasr** [comment]
Let's use 3.11-3.14 so that we have two free-threading releases. Let's just pick rmm to simplify the discussion. I would build one abi3 wheel for the non free-threaded builds that works on all Python>=3.11, and then I would build two free-threaded wheels, one for 3.13 and 3.14. When 3.15 is released, I would add a new 3.15 free-threaded build, with no change to the not free-threaded build.

Is that what you were looking for in a proposal?

Of course, there are plenty of open questions to address:
- Free threading could get rolled back since the acceptance was provisional
- If it doesn't get rolled back, [we shouldn't expect two separate builds to exist in perpetuity](https://discuss.python.org/t/pep-703-making-the-global-interpreter-lock-optional-in-cpython-acceptance/37075)

> We have to assume the free-threaded build will be the only build in the reasonable future

- I haven't tracked discussions on what the plan is for the limited API in the final version of the free threaded build of Python

**Comment by jakirkham** [comment]
Thanks Vyas! 🙏

That's a good start

Now for the package using the limited API/stable ABI, how will the pack metadata capture their intended compatibility? IOW what would go in `pyproject.toml`?

**Comment by vyasr** [comment]
I don't think anything changes in pyproject.toml. It's still the same package. When you build the wheel you have to specify the appropriate flags and it changes the output filename. I don't know which build backends are currently supporting the wheel name appropriately (since we have to add the `t` suffix), so we may have to tag wheels separately. For conda packages I think we would use a separate label until conda offers something different (labels are what [Stan's post here implicitly suggests](https://community.anaconda.cloud/t/try-free-threading-the-future-of-parallel-processing-in-python/83855)).

**Comment by jameslamb** [comment]
> *you have to specify the appropriate flags and it changes the output filename*

Just want to mention PyPA has a tool called `abi3audit` (https://github.com/pypa/abi3audit). It'd be helpful in this effort... it doesn't have a `repair` command similar to `auditwheel`, but it could at least be used to raise errors in CI if we try to tag a wheel as following the stable API which actually isn't.

The "Motivation" section of the docs explains this well: https://github.com/pypa/abi3audit?tab=readme-ov-file#motivation

Context for creation of that tool: https://github.com/pypa/auditwheel/issues/395

**Comment by vyasr** [comment]
You can see an example of using abi3audit in [this pynvjitlink PR](https://github.com/rapidsai/pynvjitlink/pull/108/files).

**Comment by vyasr** [comment]
I just tested this again to see where we stood. The current rmm code with Cython 3.0.12 has some issues with limited API compilation again. When I updated to Cython 3.1.1, there were still issues, but they were new ones. It looks like there's a decent bit in flux with limited API support in recent releases, so we might have to wait until things stabilize a little bit before we can reliably test again.

**Comment by da-woods** [comment]
> The current rmm code with Cython 3.0.12 has some issues with limited API compilation again

I don't think there's much (usable) Limited API support in any 3.0.x Cython release. 

> When I updated to Cython 3.1.1, there were still issues, but they were new ones

In principle the Limited API should work in this release - it passes the majority of our own test suite for example. (There a still a few pending bug fixes for 3.1 but they're mostly not Limited API). So do please report any Limited API issues with this release.

**Comment by vyasr** [comment]
With https://github.com/rapidsai/cudf/pull/18974 and https://github.com/rapidsai/raft/pull/2686 and a couple of fixes to Cython master that have been merged in the last 48 hours, I think all of RAPIDS will successfully at least build using the limited API. That says nothing about whether tests will pass, but that is still a significant milestone if we can hit it. Once that works we can have a discussion about what a suitable rollout of testing could look like for us to build confidence in performance and usability.

**Comment by vyasr** [comment]
As of today I am able to build all of RAPIDS other than ucxx under the limited API locally using Cython's master branch. I also ran rmm and cudf's test suites, and both passed with no meaningful performance regressions:
- rmm (no parallelism due to OOMs)
	- With: 2:44
	- Without: 2:46
- cudf (-n 16):
	- With: 5:53
	- Without: 5:45

However, I was building and testing in the same environment, so there may be uncaught bugs that have to do with building within e.g. a 3.13 environment and running in a 3.11 environment (I was targeting the 3.11 limited API, though, since that is the minimum that we can use since that is where the buffer protocol was added to the limited API). The best way to test that is probably just opening PRs in RAPIDS repositories. In the meantime, I pushed up the necessary changes for testing in https://github.com/rapidsai/devcontainers/pull/278 so if CI passes there that'll be substantial progress here.

**Comment by vyasr** [comment]
We should consider the possibility that some of our packages will build and test successfully under the limited API while others do not, and if that happens whether we are willing to move some packages over to using it before others (analogous to https://github.com/rapidsai/build-planning/issues/174#issuecomment-2858970471 for free-threading).

**Comment by vyasr** [comment]
The free-threaded build of Python 3.15 [will support the stable ABI](https://discuss.python.org/t/pep-779-criteria-for-supported-status-for-free-threaded-python/84319/123), and [a PEP is now open to track that work](https://peps.python.org/pep-0803/), so aside from a brief interim period there shouldn't be any conflict between our moving to limited API builds and supporting free threading. Depending on whether or not we get free threading support in our Python 3.14 builds or if we don't get there until 3.15, we may not even have to worry about bifurcating builds temporarily for that support.

**Comment by gforsyth** [comment]
I put up a PR in https://github.com/rapidsai/gha-tools/pull/233 adding a `--stable` flag to `rapids-package-name` to allow us to name artifacts built with the limited API in a predictable way

**Comment by gforsyth** [comment]
Ok, here's where we're at:

* [x] Support for stable ABI builds merged into `rapids-cmake` (https://github.com/rapidsai/rapids-cmake/pull/954)
* [x] Changes to `rapids-package-name` for stable ABI builds (https://github.com/rapidsai/gha-tools/pull/233)
* [x] RMM PR demonstrating stable abi builds passing tests (https://github.com/rapidsai/rmm/pull/2184)

The minimum CPython stable ABI we can work with is 3.11, because that's when the buffer protocol was added.

The easiest path forward from that is to simply drop support for Python 3.10.
We're waiting to hear from @jacobtomlinson about whether some of the major cloud providers are still running their various Python services on top of 3.10 which would complicate dropping 3.10 support.

If we can drop 3.10, we'll want to have a deprecation warning in 26.02 before dropping it in 26.04.
If we can't, we'll have to bifurcate the builds and keep 3.10 builds as they currently are, while later Python versions can use the stable abi builds.

**Comment by jacobtomlinson** [comment]
Python 3.10 is not end-of-life until October 2026, cloud platforms often will only upgrade when they have to, either due to lack of security fixes or strong customer pressure. However, it seems Snowflake Notebooks are the only major cloud notebook service still running on Python 3.10 with no user ability to upgrade this. The consequence of dropping 3.10 would be that they couldn't upgrade their RAPIDS versions to the latest and greatest until they also upgraded their Python version.

It feels like a product decision whether we need to ensure they can install the latest RAPIDS, or whether we should go ahead and just encourage them to upgrade.

**Comment by gforsyth** [comment]
I tried out the build splitting in the RMM PR.  It's not terrible, but I agree it's a product decision.
https://github.com/rapidsai/rmm/pull/2184

**Comment by gforsyth** [comment]
Trying to see how much of RAPIDS we can build (with passing tests) using the limited API:

* [x] rmm (https://github.com/rapidsai/rmm/pull/2184)
* [x] kvikio (https://github.com/rapidsai/kvikio/pull/892)
* [x] cudf (https://github.com/rapidsai/cudf/pull/20974)
* [x] ucxx (https://github.com/rapidsai/ucxx/pull/564)
* [x] raft (https://github.com/rapidsai/raft/pull/2915)
* [x] cugraph (https://github.com/rapidsai/cugraph/pull/5388)
* [x] cuvs (https://github.com/rapidsai/cuvs/pull/1687)
* [x] rapidsmpf (https://github.com/rapidsai/rapidsmpf/pull/773)
* [x] cuml (https://github.com/rapidsai/cuml/pull/7659)
* [x] cugraph-gnn (https://github.com/rapidsai/cugraph-gnn/pull/383)
* [ ] cuforest (https://github.com/rapidsai/cuforest/pull/28)
* [x] nx-cugraph (test only) (https://github.com/rapidsai/nx-cugraph/pull/225)
* [x] cuxfilter (test only) (https://github.com/rapidsai/cuxfilter/pull/754)

**Comment by vyasr** [comment]
The various PRs linked in [the above comment](https://github.com/rapidsai/build-planning/issues/42#issuecomment-3719666880) demonstrate that we can build all our packages using the limited API and that tests pass, which is great progress. The last major step here is for us to run benchmarks to ensure that using the limited API does not have meaningful performance impacts on our libraries. For whoever helps us with benchmarking, the steps are:

- Create a new environment for benchmarking. The easiest option is probably installation via pip, so a Python venv will work.
- Run the per-project scripts available in https://github.com/gforsyth/test_limited_API_wheels to download all of the experimental wheels and install them in the activate environment
- Run a representative set of benchmarks. I recommend two sets of benchmarks:
    - A set of the typical benchmarks that you run for your project. For example, TPC-H benchmarks for cudf, or scikit-learn benchmarks for cuml
    - Some microbenchmarks that will stress smaller API calls. The idea is to find code paths that perform the highest number of CPython function calls for a given operation, rather than spending a lot of time in CUDA kernels. The purpose here is to gauge the worst-case performance.
- Rerun the benchmarks on a vanilla (non-limited API) build for comparison

**Comment by gforsyth** [comment]
Putting up PRs to also build limited API packages with `rattler-build` -- those PRs aren't interconnected like the wheel ones are.  Assuming the benchmarking tests don't flag up anything too critical, I can combine the wheel and conda PRs easily.

* [ ] rmm (https://github.com/rapidsai/rmm/pull/2221)
* [ ] kvikio (https://github.com/rapidsai/kvikio/pull/905)

**Comment by gforsyth** [comment]
We have PIC signoff (or at least reports of no notable performance regressions) for:

- rmm
- cuml
- cudf

PRs for full switchover to limited API builds will be listed out here:
- [x] https://github.com/rapidsai/rmm/pull/2241
- [x] https://github.com/rapidsai/kvikio/pull/920
- [x] https://github.com/rapidsai/cuml/pull/7758
- [x] https://github.com/rapidsai/cudf/pull/21329
- [x] https://github.com/rapidsai/raft/pull/2943
- [x] https://github.com/rapidsai/cugraph/pull/5419
- [x] https://github.com/rapidsai/cuvs/pull/1788
- [x] https://github.com/rapidsai/cugraph-gnn/pull/407 (not currently compatible with limited API -- uses `array.array`)
- [x] https://github.com/rapidsai/cuforest/pull/43

Unblocked:
- [x] https://github.com/rapidsai/rapidsmpf/pull/850 (segfaults because of a Cython bug that is fixed on Cython `main` but not yet released)

With https://github.com/rapidsai/rapidsmpf/pull/878 there are no more nested closures in `rapidsmpf` (which was causing the Cython issue) -- once that is merged in, limited API builds should work for `rapidsmpf`

Blocked:

- [ ] https://github.com/rapidsai/cucim/pull/1027 (uses `pybind11` which doesn't support the limited API. Devs looking at potentially switching to `nanobind`)

**Comment by gforsyth** [comment]
@wence- raised in https://github.com/rapidsai/rapidsmpf/pull/850 that with limited API builds in CI only, there's now a mismatch between CI and local developer environments.

To resolve that, I'm planning to:

- [ ] ~open a PR to `rapidsai/devcontainers` to default to buliding Python packages with the appropriate limited API args~
- [ ] ~rapids-reviser PR to update `build.sh` to have a default value for `RAPIDS_PY_API` so that devs using `build.sh` also get limited API builds~

Passing in `wheel.py-api` via command line overrides the contents of the `pyproject.toml`, so I think the lowest friction way to handle this is to hard-code the current `py-api` (`cp311`) into the `pyproject.toml` files and then update them as needed (but without _needing_ to update them for us to bump the lower bound for published packages)

- [x] https://github.com/rapidsai/rmm/pull/2263
- [x] https://github.com/rapidsai/kvikio/pull/934
- [x] https://github.com/rapidsai/cudf/pull/21560
- [x] https://github.com/rapidsai/raft/pull/2967
- [x] https://github.com/rapidsai/rapidsmpf/pull/887
- [x] https://github.com/rapidsai/cuvs/pull/1852
- [x] https://github.com/rapidsai/cuml/pull/7832
- [x] https://github.com/rapidsai/cugraph/pull/5440
- [x] https://github.com/rapidsai/cugraph-gnn/pull/415
- [x] https://github.com/rapidsai/cuforest/pull/50

**Comment by da-woods** [comment]
Also bear in mind that you occasionally get a slight mismatch depending on what version of the Python headers you compile with - e.g. if you compile against the Python 3.14 headers with `Py_LIMITED_API=0x030B0000` it isn't always quite the same as compiling against the 3.11 headers. It's generally fine and the Python developers have got a lot more practised at getting it right in recent releases. But occasionally there's debugging fun to be had there.

**Comment by jakirkham** [comment]
@gforsyth can see that our Conda packages build with Python 3.11. However it looks like some wheel builds use Python 3.13 (for [example]( https://github.com/rapidsai/cugraph/actions/runs/21831924241/job/62992782909#step:13:68 )). Would it be possible to make sure both build with Python 3.11?

That way we can avoid the subtle issues that David has alluded to above

**Comment by gforsyth** [comment]
Hey @jakirkham -- C++ wheels are built with the latest Python we support (currently 3.13, soon to be 3.14), but all Python wheels that use the limited API are built with Python 3.11 (excepting `cucim` and `ucxx` which still built a wheel-per-version)

**Comment by jakirkham** [comment]
Is there a reason not to use the same Python-versioned image for all wheel builds?

**Comment by gforsyth** [comment]
The C++ wheels should be inherently agnostic to the Python version used to build them.

So there isn't a reason to not use the same Python-versioned image, but also using different ones also doesn't really have an impact.

We can certainly change the C++ wheel builds to use Python 3.11 images, just not sure where it falls w.r.t priority

**Comment by jakirkham** [comment]
What is entailed in changing the version? Just that or does this touch some script somewhere used for managing CI jobs as well?

**Comment by gforsyth** [comment]
> What is entailed in changing the version? Just that or does this touch some script somewhere used for managing CI jobs as well?

I think it would entail updating the `matrix_filter` in each `pr.yaml` and `build.yaml` for the C++ wheels jobs -- I don't think there's anything else involved

**Comment by gforsyth** [comment]
Ok, I think all of the initial work on this is done, and now all of the follow-up PRs to add the explicit `wheel.py-api` line to `pyproject.toml` files is also done.

`ucxx` remains to be investigated and ported over, but do we want to close out this tracking issue?

**Comment by jameslamb** [comment]
> _ucxx remains to be investigated and ported over, but do we want to close out this tracking issue?_

`ucxx` support is now merged: https://github.com/rapidsai/ucxx/pull/564

@gforsyth it would be good to close these earlier draft PRs like https://github.com/rapidsai/cugraph-gnn/pull/383 now that we don't need them.

But otherwise yes, I think this done. I'll let you click the button, thanks for making this happen!!! It makes soooo many things better 😁

**Comment by gforsyth** [comment]
Ok, all the remaining draft and testing PRs are closed now -- going to close this out as done!  (We're still not building `cucim` with the limited API but moving that over to use Cython or `nanobind` instead of `pybind11` is out of scope for this.

That work is tracked in https://github.com/rapidsai/cucim/issues/1029

**Comment by jameslamb** [comment]
While going to add `abi3audit` to CI in RAPIDS, I realized that `cuopt` wasn't included in the Limited API effort.

I'm not sure if that was intentional or just an oversight... I don't see any discussion about it here.

Put up a tracking issue over here: https://github.com/NVIDIA/cuopt/issues/1725
